modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -1317,152 +1379,153 @@ private RuntimeException convertThrowable(Throwable th) {
-     * Ignite#setBaseline(Set)}.
-     *
-     * @param clusterNodes Set of nodes for assignment.
-     */
-    private void updateAssignments(Set<ClusterNode> clusterNodes) {
-        var setBaselineFut = new CompletableFuture<>();
 
-        var changePeersQueue = new ArrayList<Supplier<CompletableFuture<Void>>>();
+                List<List<ClusterNode>> assignments = (List<List<ClusterNode>>)
+                        ByteUtils.fromBytes(((ExtendedTableConfiguration) tablesCfg.tables().get(tbl.name())).assignments().value());
 
-        tablesCfg.tables()
-                .change(tbls -> {
-                    changePeersQueue.clear();
+                List<ClusterNode> newPeers = ((List<ClusterNode>) ByteUtils.fromBytes(evt.entryEvent().newEntry().value()));
 
-                    for (int i = 0; i < tbls.size(); i++) {
-                        tbls.createOrUpdate(tbls.get(i).name(), changeX -> {
-                            ExtendedTableChange change = (ExtendedTableChange) changeX;
-                            byte[] currAssignments = change.assignments();
+                RaftGroupService raftGrpSvc = null;
 
-                            List<List<ClusterNode>> recalculatedAssignments = AffinityUtils.calculateAssignments(
-                                    clusterNodes,
-                                    change.partitions(),
-                                    change.replicas());
+                try {
+                    var deltaPeers = newPeers.stream()
+                            .filter(p -> !assignments.get(part).contains(p))
+                            .collect(Collectors.toList());
 
-                            if (!recalculatedAssignments.equals(ByteUtils.fromBytes(currAssignments))) {
-                                change.changeAssignments(ByteUtils.toBytes(recalculatedAssignments));
+                    raftGrpSvc = raftMgr.updateRaftGroup(grpId, assignments.get(part), deltaPeers, raftGrpLsnrSupplier,
+                            raftGrpEvtsLsnrSupplier).join();
 
-                                changePeersQueue.add(() ->
-                                        updateRaftTopology(
-                                                (List<List<ClusterNode>>) ByteUtils.fromBytes(currAssignments),
-                                                recalculatedAssignments,
-                                                change.id()));
-                            }
-                        });
-                    }
-                })
-                .thenCompose((v) -> {
-                    CompletableFuture<?>[] changePeersFutures = new CompletableFuture<?>[changePeersQueue.size()];
+                    raftGrpSvc.refreshLeader().join();
 
-                    int i = 0;
+                    if (new Peer(raftGrpSvc.clusterService().topologyService().localMember().address()).equals(raftGrpSvc.leader())) {
```

Comment:
How do we can sure that the leader won't change just after?

Reply:
> How do we can sure that the leader won't change just after?

It's actually not a problem. In worst case we will have attempt to start several rebalances to the same set of target peers, because each onLeaderElected() will also start one. @kgusakov As far as I remember we've planned to propagate term to chagnePeers to skip such not-a-leader anymore rebalances.